src/history/hs.c

```diff
@@ -79,22 +118,21 @@ __wt_hs_config(WT_SESSION_IMPL *session, const char **cfg)
     if (F_ISSET(conn, WT_CONN_IN_MEMORY | WT_CONN_READONLY))
         return (0);
 
+    WT_ERR(__hs_start_internal_setup_session(session, &temp_setup_session));
+    release_temp_session = true;
+
     /*
-     * Retrieve the btree from the history store cursor. This function might need to open a cursor
-     * from the default session and hence need to flip the no-dhandle flag temporarily in case it is
-     * set. TODO: WT-5501 We should find a way to do this without opening a history store cursor
-     * from the default session.
+     * Retrieve the btree from the history store cursor.
      */
-    if (F_ISSET(session, WT_SESSION_NO_DATA_HANDLES)) {
-        reset_no_dh_flag = true;
-        F_CLR(session, WT_SESSION_NO_DATA_HANDLES);
-    }
-    WT_ERR(__wt_hs_get_btree(session, &btree));
+    WT_ERR(__wt_hs_get_btree(temp_setup_session, &btree));
 
     /* Track the history store file ID. */
     if (conn->cache->hs_fileid == 0)
         conn->cache->hs_fileid = btree->id;
 
+    /* Cache the history store btree reference. */
+    conn->hs_btree = btree;
```

Comment:
Is it always OK to stash this btree? I see that the temporary session is closed further down here - is it possible that the dhandle and it's associated btree will be cleaned up before another session starts using them? That seems unlikely - I think the sweep server would need to come through in a very small window generally, but it seems theoretically possible.

Reply:
I am concerned. I think it can be opened here, and the reference released (when the temporary session is closed). Then sweep is enabled and can possibly close the dhandle/btree. Then another session can use the stashed dhandle/btree.

Reply:
Will that make this code hazardous as well? We are also stashing the `fileID` for cache.
`/* Track the history store file ID. */
    if (conn->cache->hs_fileid == 0)
        conn->cache->hs_fileid = btree->id;`

Reply:
That's a good question. The file ID doesn't change - it's used to ensure that logging applies records to the correct table, so it must be the same for the lifetime of a table. Which means it's safe to save it here.

Reply:
@agorrod Can the dhandle sweep ignore the history store? We're going to be constantly opening it so my thoughts are that it should be fine to just omit it from the sweep.

Reply:
I thought about that as well, but I'd rather avoid exceptional handling for the history store where possible.
I talked to Haseeb, and I think we can work around this problem. There is no need to save the btree handle in the connection at the moment.